File: src/main/java/com/joraph/JoraphContext.java
package com.joraph;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.RejectedExecutionHandler;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import com.joraph.loader.EntityLoaderContext;
import com.joraph.schema.Schema;

/**
 * The main class for using Joraph.
 */
public class JoraphContext {

	private Schema schema;
	private EntityLoaderContext loaderContext;
	private ExecutorService executorService;
	private long parallelExecutorDefaultTimeoutMillis = TimeUnit.SECONDS.toMillis(30);

	/**
	 * Creates a context for the given {@link Schema}.
	 * @param schema
	 */
	public JoraphContext(Schema schema) {
		this(schema, 50);
	}

	/**
	 * Creates a context for the given {@link Schema}.
	 * @param schema
	 * @param parallelExecutorCount
	 */
	public JoraphContext(Schema schema, int parallelExecutorCount) {
		this.schema 		= schema;
		this.loaderContext	= new EntityLoaderContext();
		setParallelExecutorCount(parallelExecutorCount);
	}

	/**
	 * Returns the configurd {@link EntityLoaderContext}.
	 * @return the {@link EntityLoaderContext}
	 */
	public EntityLoaderContext getLoaderContext() {
		return loaderContext;
	}

	/**
	 *
	 */
	public ObjectGraph createEmptyGraph() {
		return new ObjectGraph(schema);
	}

	/**
	 *
	 */
	public ObjectGraph execute(Query query) {
		return new ExecutionContext(this, query).execute();
	}

	/**
	 *
	 */
	public <T> ObjectGraph execute(Class<T> entityClass, Collection<T> objects) {
		return execute(new Query()
				.withEntityClass(entityClass)
				.withRootObjects(objects));
	}

	/**
	 *
	 */
	public ObjectGraph execute(Collection<Class<?>> entityClasses, Collection<Object> objects) {
		return execute(new Query()
				.withEntityClasses(entityClasses)
				.withRootObjects(objects));
	}

	/**
	 *
	 */
	public <T> ObjectGraph execute(Class<T> entityClass, Collection<T> objects, ObjectGraph existingGraph) {
		return execute(new Query()
			.withEntityClass(entityClass)
			.withRootObjects(objects)
			.withExistingGraph(existingGraph));
	}

	/**
	 *
	 */
	public ObjectGraph execute(Collection<Class<?>> entityClasses, Collection<Object> objects, ObjectGraph existingGraph) {
		return execute(new Query()
			.withEntityClasses(entityClasses)
			.withRootObjects(objects)
			.withExistingGraph(existingGraph));
	}

	/**
	 *
	 */
	public <T> ObjectGraph execute(Class<T> entityClass, T rootObject) {
		return execute(new Query()
			.withEntityClass(entityClass)
			.withRootObjects(CollectionUtil.asSet(rootObject)));
	}

	/**
	 *
	 */
	public <T> ObjectGraph execute(Class<T> entityClass, T rootObject, ObjectGraph existingGraph) {
		return execute(new Query()
			.withEntityClass(entityClass)
			.withRootObjects(CollectionUtil.asSet(rootObject))
			.withExistingGraph(existingGraph));
	}

	/**
	 *
	 */
	public ObjectGraph executeForRootObject(Object rootObject) {
		assert(rootObject != null);
		final Class<?> entityClass = rootObject.getClass();
		assert(entityClass != null);
		
		return execute(new Query()
			.withEntityClass(entityClass)
			.withRootObjects(CollectionUtil.asSet(rootObject)));
	}

	/**
	 *
	 */
	public ObjectGraph executeForRootObject(Object rootObject, ObjectGraph existingGraph) {
		assert(rootObject != null);
		final Class<?> entityClass = rootObject.getClass();
		assert(entityClass != null);
	
		return execute(new Query()
			.withEntityClass(entityClass)
			.withRootObjects(CollectionUtil.asSet(rootObject))
			.withExistingGraph(existingGraph));
	}

	/**
	 *
	 */
	public <T> List<T> load(Class<T> entityClass, Collection<Object> ids) {
		return loaderContext.load(entityClass, ids);
	}

	/**
	 * Supplements an existing graph with the given data.
	 */
	public <T> ObjectGraph supplement(
			ObjectGraph existingGraph, Class<?> entityType, Collection<Object> ids) {

		List<?> objects = load(entityType, ids);
		if (objects==null || objects.isEmpty()) {
			return existingGraph;
		}

		return execute(new Query()
				.withEntityClass(entityType)
				.withRootObjects(objects.stream()
						.map(Object.class::cast)
						.collect(Collectors.toList()))
				.withExistingGraph(existingGraph));
	}

	public Schema getSchema() {
		return schema;
	}

	public ExecutorService getExecutorService() {
		return executorService;
	}

	public void setParallelExecutorCount(int parallelExecutorCount) {
		ThreadPoolExecutor executorService = new ThreadPoolExecutor(
				parallelExecutorCount, parallelExecutorCount,
                0L, TimeUnit.MILLISECONDS,
                new LinkedBlockingQueue<Runnable>());
		executorService.setRejectedExecutionHandler(new RejectedExecutionHandler() {
			@Override
			public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
				r.run();
			}
		});
		this.executorService = executorService;
	}

	public long getParallelExecutorDefaultTimeoutMillis() {
		return parallelExecutorDefaultTimeoutMillis;
	}

	public void setParallelExecutorDefaultTimeoutMillis(
			long parallelExecutorDefaultTimeoutMillis) {
		this.parallelExecutorDefaultTimeoutMillis = parallelExecutorDefaultTimeoutMillis;
	}

}

File: src/main/kotlin/com/joraph/JoraphContext.kt
package com.joraph

import com.joraph.loader.EntityLoaderContext
import com.joraph.schema.Schema
import java.util.concurrent.*

/**
 * The main point of contact for using Joraph. The [JoraphContext] maintains a [Schema] and a
 * [EntityLoaderContext] and creates [ExecutionContext]s for executing [Query]s through it's
 * [execute] method.
 */
open class JoraphContext @JvmOverloads constructor(
    open val schema: Schema,
    parallelExecutorCount: Int = 50) {

    val loaderContext: EntityLoaderContext = EntityLoaderContext()
    var executorService: ExecutorService
    var parallelExecutorDefaultTimeoutMillis = TimeUnit.SECONDS.toMillis(30)

    init {
        val executorService = ThreadPoolExecutor(
            parallelExecutorCount, parallelExecutorCount,
            0L, TimeUnit.MILLISECONDS,
            LinkedBlockingQueue()
        )
        executorService.rejectedExecutionHandler = RejectedExecutionHandler { r, _ -> r.run() }
        this.executorService = executorService
    }

    /**
     * Creates a new [ObjectGraph] configured with the context's [Schema]
     */
    open fun createEmptyGraph(): ObjectGraph {
        return ObjectGraph(schema)
    }

    /**
     * Executes the given [Query] and returns the resulting [ObjectGraph].
     */
    open fun execute(query: Query): ObjectGraph {
        return ExecutionContext(this, query).execute()
    }

    /**
     * A shortcut for executing a query for the given [entityClasses] and root [objects].
     */
    @JvmOverloads
    open fun execute(entityClasses: Collection<Class<*>>, objects: Collection<Any>, existingGraph: ObjectGraph? = null): ObjectGraph {
        return execute(Query()
            .withEntityClasses(entityClasses)
            .withRootObjects(objects)
            .withExistingGraph(existingGraph))
    }

    /**
     * A shortcut for executing a query for the given [entityClass] and root [objects].
     */
    @JvmOverloads
    open fun execute(entityClass: Class<*>, objects: Collection<Any>, existingGraph: ObjectGraph? = null): ObjectGraph {
        return execute(Query()
            .withEntityClass(entityClass)
            .withRootObjects(objects)
            .withExistingGraph(existingGraph))
    }

    /**
     * A shortcut for executing a query for the given [entityClass] and root [rootObject].
     */
    @JvmOverloads
    open fun execute(entityClass: Class<*>, rootObject: Any, existingGraph: ObjectGraph? = null): ObjectGraph {
        return execute(Query()
            .withEntityClass(entityClass)
            .withRootObject(rootObject)
            .withExistingGraph(existingGraph))
    }

    /**
     * A shortcut for executing a query for the given [rootObject].
     */
    @JvmOverloads
    open fun executeForRootObject(rootObject: Any, existingGraph: ObjectGraph? = null): ObjectGraph {
        return execute(Query()
            .withEntityClass(rootObject.javaClass)
            .withRootObject(rootObject)
            .withExistingGraph(existingGraph))
    }

    /**
     * Allows for loading entities of the type defined by [entityClass] by their keys [ids]
     * using the configured [EntityLoaderContext].
     */
    open fun <T> load(entityClass: Class<T>, ids: Collection<Any>): List<T> {
        return loaderContext.load(entityClass, ids)
    }

    /**
     * Loads entities of the type defined by [entityType] with the primary keys [ids]
     * into the [existingGraph].
     */
    open fun <T> supplement(
        existingGraph: ObjectGraph, entityType: Class<*>, ids: Collection<Any>): ObjectGraph {
        val objects = load(entityType, ids)
        return if (objects.isEmpty()) {
            existingGraph
        } else {
            execute(Query()
                .withEntityClass(entityType)
                .withRootObjects(objects)
                .withExistingGraph(existingGraph))
        }
    }

}

